reactioncommerce/reaction
# feat: Translations without Meteor

Impact: **minor**  
Type: **feature**

## Issue
Reaction Admin i18next loads translations from a Translations collection, loaded by a Meteor DDP subscription. We are trying to remove the Meteor dependency from the API. 

## Solution
- The API (specifically the `i18n` core plugin) now provides translations, merged from all plugins, at the standard `/locales/resources.json` route.
- Because any plugin can add a translation namespace, the API (specifically the `i18n` core plugin) also provides a `/locales/namespaces.json` route, which returns the namespaces array. The i18next client library needs this list for initialization.
- In browser code, Reaction Admin i18next lib now loads translations in a standard way, per their docs.
- In plugins, translation files are unchanged but have been moved into the `no-meteor` folder and no longer use `loadTranslations` function. Instead, they pass translations array to `registerPlugin` within an `i18n` object.
- The API no longer loads translations from `/private` (this was a Meteor thing). Those translations have been moved into `core` plugin.
- The API no longer saves translations to the `Translations` collection on startup, no longer publishes that collection, and no longer supports "flushing" translations because they are now always automatically up to date.
- A migration removes all documents from the `Translations` collection.

## Other changes
- A more limited and standard language detection is now used by Reaction Admin. It works like this:
  - You will see UI text in whatever language is set as your browser default. For example, in Chrome, go to chrome://settings/languages and whatever language is listed first in your preferences list will be the language used.
  - To see the UI in a specific language, overriding your browser language, you can append `lng=<code>` in your query string. For example, `http://localhost:3000/operator/?lng=en`
 or `http://localhost:3000/operator/?lng=en-US` will show you English text even if your browser language is set to something else.
  - If some bit of text does not have a translation for your language, you will see it in English. (Consider submitting a pull request to add the missing translation!)
- The `shop/getLocal` Meteor method is removed.

## Breaking changes
- Any custom plugins relying on translations being in the Translations collection will need to be updated.
- Any custom plugins importing and using `loadTranslations` to load translations will need to be updated to pass them to `registerPlugin` instead. See the built-in plugins for examples. 
- Any custom plugins calling `shop/getLocale` Meteor method or using `Reaction.Locale` object will need changes. `i18next.language` should be used to get the detected/default language.

## Testing
1. Change your browser default/first preferred language to any language Reaction supports. For example, in Chrome, go to chrome://settings/languages and edit your list.
1. Go to http://localhost:3000/operator/ and verify that text is shown in that language (other than places where there is no translation for that language, which will be English)
1. Go to http://localhost:3000/operator/?lng=en (or any language other than your browser default) and verify that text shows in that language now.
1. Click through all possible paths in the Reaction Admin UI and make sure no text is missing or showing only the translation key rather than the text.
